### ISSUE TITLE
Envoy/C-Ares bug can cause istio to become "stuck" to one DNS pod

### ISSUE THREAD

**Issue by rjbez17** [istio/istio#53577]
### Is this the right place to submit this?

- [X] This is not a security vulnerability or a crashing bug
- [X] This is not a question about how to use Istio

### Bug Description

This is half bug report and half feature request. I've been unable to find any current configuration options to prevent this. 

Due to the way c-ares and envoy do UDP DNS requests, there are cases in which an [envoy pod will bypass the kube-dns service and send all requests to a single pod](https://github.com/envoyproxy/envoy/issues/7965). Unfortunately, the way this manifests is that many if not all istio pods running envoy get pinned to the same service. We've seen this start happening in some of our larger clusters with many istio ingress gateways and sidecar proxies (with `ISTIO_META_DNS_CAPTURE` set to `true`).

[c-ares and envoy have exposed a `udp_max_queries` configuration](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/network/dns_resolver/cares/v3/cares_dns_resolver.proto#extensions-network-dns-resolver-cares-v3-caresdnsresolverconfig) option to prevent too many requests going into a single connection. The feature request here is to either set this to a sane value or expose it to be configured by end users. Any other suggestion to prevent this issue would also be welcomed. 

### Version

```Text
This affects all istio versions
```

### Additional Information

The screen shot below shows this manifesting in our cluster, you can see nearly all requests routed to a single DNS pod and the traffic to all other CoreDNS pods drop to near zero. 
![Screenshot 2024-10-15 at 3 14 12 PM](https://github.com/user-attachments/assets/8bba6f2e-fad9-4b6a-8d9d-1af15dfeca8e)

**Comment by howardjohn** [comment]
cc @Stevenjin8

**Comment by Stevenjin8** [comment]
@howardjohn Meshconfig seems like a good way to expose this? Still not super familiar with the different ways to configure istio.

**Comment by howardjohn** [comment]
Having a default value here seems reasonable as well. Whether we need customization I am not even sure

**Comment by Stevenjin8** [comment]
@rjbez17 Do you have any thoughts on default value vs having it be configurable? What would be a reasonable default value? I have no context for this and looking around, it doesn't seem like many people this.

**Comment by ramaraochavali** [comment]
FWIW, we set it to 10 and it has fixed the issue for us.  You can also set `use_resolvers_as_fallback` so that Envoy accepts empty resolvers. You do not need the other PR in Envoy

```
"resolvers": [],
 "use_resolvers_as_fallback": true,
```

**Comment by rjbez17** [comment]
That seems like a reasonable start to me yes.

**Comment by Stevenjin8** [comment]
@howardjohn Thoughts on adding this to be configurable and a default value of 10?

**Comment by howardjohn** [comment]
Hard to answer without knowing the performance impact of switching to the new 'connection'. 10 seems awfully low but its really a toss-up without that data

**Comment by rjbez17** [comment]
If I'm understanding the underlying c-ares bug, normal operating mode has new connections for every request (pretty basic for UDP traffic). The bug here is that the connection can accidentally be "left open" from an iptables perspective and then can never fully close. 

I am curious if this will cause any performance impact but from everything I understand now, its *shouldn't*.

**Comment by hzxuzhonghu** [comment]
does tcp dns requeries face same problem?

**Comment by Stevenjin8** [comment]
> does tcp dns requeries face same problem?

Not sure, but I don't see a configuration for tcp

https://c-ares.org/docs/ares_init.html

**Comment by ramaraochavali** [comment]
for tcp it is not an issue AFAIK but tcp has others scalability issues

**Comment by hzxuzhonghu** [comment]
@ramaraochavali Can you elaborate a little bit more about the scalability

**Comment by ramaraochavali** [comment]
https://www.geeksforgeeks.org/why-does-dns-use-udp-and-not-tcp/

**Comment by Stevenjin8** [comment]
The discussion is spread across three multiple PR/issues, but the main point is that we strongly prefer having a reasonable default than expanding the API surface. After some discussion in https://github.com/c-ares/c-ares/issues/913, the main conclusion is that the only way we can regress performance is by setting `udp_max_queries` to be too low. This is because it is possible for every available UDP port to have maxed out the number of queries, but still be awaiting for a response. However, this regression can only happen if we are performing DNS queries at a crazy rate. Assuming that dns queries are being slow (5s, the default cares timeout), we have `udp_max_queries=10` and 10000 local udp ports available, we would have perform 10 * 10000 / 5 = 20000 queries per second to exhaust our ports.

Even if we tried to sabotage ourselves with `udp_max_queries=1`, we would still need to be performing 2000 queries per second. In the above example, **dns pods** are only receiving 15000/110=1363 requests per second. So we can assume that each envoy is sending much fewer.

**The point is 10 seems like a reasonable value**.

**Comment by howardjohn** [comment]
20k QPS is unfortunately on the low end. We have seen FAR higher. With ServiceEntries, the formula is typically `# SE * # Proxy / TTL`.

Lets say we have 1000 SE and 1000 proxies (**not** a large cluster) with a typical 30s TTL -- that is 500k QPS!! It gets worse when you factor in search domains can make this ~5x higher. This is also why we need https://github.com/envoyproxy/envoy/issues/20562 btw.

Given we are going from ~infinite, I would error on the side of too high than too low and go with something like 1000?

**Comment by Stevenjin8** [comment]
> 20k QPS is unfortunately on the low end. We have seen FAR higher. With ServiceEntries, the formula is typically `# SE * # Proxy / TTL`.
> 
> Lets say we have 1000 SE and 1000 proxies (**not** a large cluster) with a typical 30s TTL -- that is 500k QPS!! It gets worse when you factor in search domains can make this ~5x higher. This is also why we need [envoyproxy/envoy#20562](https://github.com/envoyproxy/envoy/issues/20562) btw.
> 
> Given we are going from ~infinite, I would error on the side of too high than too low and go with something like 1000?

I think the formula is # SE / TTL since this settings applies to each envoy, not the total amount of DNS queries in a cluster.

**Comment by howardjohn** [comment]
ahh right. Good point!

Ok, 100 is probably reasonable then

**Comment by JingmingGuo** [comment]
@rjbez17 We experienced the same issue. The DNS traffic metric is exactly as you link here. Has your issue been resolved by setting `udp_max_queries`?

**Comment by Stevenjin8** [comment]
@JingmingGuo I believe we resolved this in https://github.com/istio/istio/pull/53806 by setting to a reasonable value. What version of istio are you using?

**Comment by JingmingGuo** [comment]
@Stevenjin8 We are using an older version but we can take in this change to set `udp_max_queries`. It's confusing that the traffic here to one coredns pod is much higher than one envoy can send(same for us). If one envoy/c-ares stick with one connection, how could the other envoy in the cluster result in stick with the same one connection?

**Comment by Stevenjin8** [comment]
@JingmingGuo I don't super understand the problem, but I'll try to summarize it so preface everything I'm about to say with "I think". One envoy gets stuck to a coredns pod cause the coredns pod to be really so. As the other envoys cycle through coredns pods, they will eventually hit the slow one and themselves get stuck on it. Eventually all envoys are stuck on that pod.

**Comment by rjbez17** [comment]
You can read more info here: https://github.com/envoyproxy/envoy/issues/7965

But high level, if a UDP destination behind a kube service has a high enough latency, an envoy proxy can get 'stuck' to bypassing the normal routing (iptables) and send all requests to that one pod. 

Since the istio sidecar is based on envoy, all sidecars could hit this bug as the latency increases and therefore quickly you could have all DNS requests going to a single pod.

**Comment by rjbez17** [comment]
@Stevenjin8 I'm having trouble finding which versions of istio this fix was release in. Its not in the release notes of any releases after the merge in mid novemeber (that I can tell). Did this make it into the 1.22.7, 1.23.4, and 1.24.1 release?

**Comment by rjbez17** [comment]
> You can read more info here: [envoyproxy/envoy#7965](https://github.com/envoyproxy/envoy/issues/7965)
> 
> But high level, if a UDP destination behind a kube service has a high enough latency, an envoy proxy can get 'stuck' to bypassing the normal routing (iptables) and send all requests to that one pod.
> 
> Since the istio sidecar is based on envoy, all sidecars could hit this bug as the latency increases and therefore quickly you could have all DNS requests going to a single pod.

Its also worth noting we reproduced this issue with IPVS

**Comment by Stevenjin8** [comment]
@rjbez17 Sorry just saw this. Should be available in 1.25
